### PR TITLE
Device agnostic testing

### PR DESCRIPTION
**Pull request by vvvm23**
# What does this PR do?

Adds extra capabilities to `testing_utils.py` to support testing on devices besides from `cuda` and `cpu` without having to upstream device-specific changes.

This involves introducing some device agnostic functions that dispatch to specific backend functions. Users can specify new backends and backends for device agnostic functions via creating a device specification file and pointing the test suite to it using `TRANSFORMERS_TEST_DEVICE_SPEC`.

An example specification for a hypothetical CUDA device without support for `torch.cuda.empty_cache` could look like this:
```python
import torch

# !! Specify additional imports here !!

# Specify the device name (eg. 'cuda', 'cpu')
DEVICE_NAME = 'cuda2'

# Specify device-specific backends to dispatch to.
# If not specified, will fallback to 'default' in 'testing_utils.py`
MANUAL_SEED_FN = torch.cuda.manual_seed
EMPTY_CACHE_FN = None
DEVICE_COUNT_FN = torch.cuda.device_count
```

By default, we have `cpu` and `cuda` backends available, so not to affect default behaviour.

We also introduce a new decorator `@require_torch_accelerator` which can be used to specify that a test needs an accelerator (but not necessarily a CUDA one).

Crucially, these changes _should_ not change the behaviour of upstream CI runners. They aim to be as non-intrusive as possible and do not break compatibility with tests before these changes are made.

In this PR, only a subset of all tests are updated to support these new features at first. These are:
- `test_modeling_bloom` – demonstrating usage of new `@require_torch_accelerator`
- `test_modeling_codegen` – demonstrating usage of device agnostic function (`accelerator_manual-seed`)
- `test_modeling_opt` – demonstrating another device agnostic function, this time to check whether the current device supports `torch.float16`
- `test_modeling_reformer` – decorator version of the above. 

Related #25654 

TODO:
- [ ] Write some documentation on `TRANSFORMERS_TEST_DEVICE_SPEC` (once we finalise the PR)
- [ ] Additional checks and finding edge cases
- [ ] Verify this PR does indeed have no effect on the Huggingface CI runners.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [X] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@ydshieh

### CODE REVIEW

**Comment by ydshieh** [comment]
Thanks, I will  take a look @vvvm23

**Comment by ydshieh** [comment]
@vvvm23 Could you pull the latest main to your local clone and rebase your PR branch on top of your local `main`?

**Comment by vvvm23** [comment]
I rebased a few days ago, but realised I forgot to ping! Sorry @ydshieh!

**Comment by ydshieh** [comment]
No problem, I will take a look this week 🙏

File: src/transformers/utils/import_utils.py
Comment (ydshieh) [inline src/transformers/utils/import_utils.py:329]: I would rather call this `is_torch_fp16_available_on_device`.

File: src/transformers/utils/import_utils.py
Comment (ydshieh) [inline src/transformers/utils/import_utils.py:348]: same. And keep the original one untouched

File: src/transformers/testing_utils.py
Comment (ydshieh) [inline src/transformers/testing_utils.py:2137]: We can use `torch.manual_seed` for `cpu`

File: src/transformers/testing_utils.py
Comment (ydshieh) [inline src/transformers/testing_utils.py:2168]: need `f` to be f-string

File: src/transformers/testing_utils.py
Comment (ydshieh) [inline src/transformers/testing_utils.py:2164]: could fit in one line?

File: src/transformers/testing_utils.py
Comment (ydshieh) [inline src/transformers/testing_utils.py:2155]: Could be one line. Otherwise, let's use `make style` to do the work.

**Review by ydshieh — COMMENTED** [review]
Hi! Thank you for opening this PR. Overall it's good! I left a few nint comments. 

I will have to check the failing tests though.

Have you tried to run the tests with a 3rd party device (with this PR of course)?

File: src/transformers/testing_utils.py
Comment (ydshieh) [inline src/transformers/testing_utils.py:2139]: We need to guard this inside `is_torch_availabe()`. If not, set all of them to be empty dict

File: src/transformers/testing_utils.py
Comment (vvvm23) [reply src/transformers/testing_utils.py:2164]: This is a result of `make style` so I think it must be like this

File: src/transformers/testing_utils.py
Comment (vvvm23) [reply src/transformers/testing_utils.py:2155]: This is a result of `make style` so I think it must be like this

File: src/transformers/testing_utils.py
Comment (vvvm23) [reply src/transformers/testing_utils.py:2139]: might be better to guard the whole block? starting from line 2154?

Comment (ydshieh) [reply src/transformers/testing_utils.py:2139]: not sure I understand. You mean only guard the lines after 2154?

Comment (vvvm23) [reply src/transformers/testing_utils.py:2139]: Yep, guard from this line
https://github.com/huggingface/transformers/blob/c961bd633e29b98e598011643de4e166fb136dd6/src/transformers/testing_utils.py#L2154

None of the logic is necessary outside of a torch context.

Comment (ydshieh) [reply src/transformers/testing_utils.py]: The definition

> ACCELERATOR_MANUAL_SEED = {"cuda": torch.cuda.manual_seed, "cpu": torch.manual_seed, "default": torch.manual_seed}

etc. already needs to be guarded, as it uses `torch`.

The definition of `accelerator_manual_seed` etc. are imported in other files, so they can't be guarded here, otherwise import will fail.

The block in `if "TRANSFORMERS_TEST_DEVICE_SPEC" in os.environ` could be guarded.

Comment (vvvm23) [reply src/transformers/testing_utils.py]: Ah sorry, misunderstood the line you meant.

Will apply these changes~

**Comment by vvvm23** [comment]
Hey @ydshieh, just writing to let you know that @arsalanu will be picking up this PR on my behalf as I am changing jobs.

Please let me or him know if you have any further review comments 🤗

**Comment by arsalanu** [comment]
Hi, as Alex mentioned, I’ll be taking over this PR. Just wanted to check in on the status of the review, please let me know when you can if there are any comments/further changes you’d like made 🙂

**Comment by ydshieh** [comment]
Other than this, it looks good to me. It would be great to see an example run on `npu` backend/device and make sure this PR works with it.

Finally, we should update the documentation here

https://github.com/huggingface/transformers/blob/b5ca8fcd20da3ed6aa562ca926c4c8e2c56fe6a4/docs/source/en/testing.md?plain=1#L514

**Comment by arsalanu** [comment]
Hi, I've updated the PR to rename the functions, I agree that `backend_*` makes more sense here. Also updated the docs to include an explanation of using the spec file.

**Comment by arsalanu** [comment]
On the example run, what is sufficient for showing the changes are working with a different backend?

We have functional tests with these changes running on a custom backend, but we are not able to share details about the backend itself, but we could, for example, provide some CI runner outputs with backend names redacted.

**Comment by ydshieh** [comment]
> On the example run, what is sufficient for showing the changes are working with a different backend?

If you have access to a npu machine, maybe just run a test (that is ) with that device/backend , and showing the device name?
Say `test_generate_fp16` in `tests/models/opt/test_modeling_opt.py`? Otherwise, probably just showing the device name doesn't contain `cuda` or `cpu`.

**Comment by arsalanu** [comment]
Hi @ydshieh, I have a [Colab notebook](https://colab.research.google.com/drive/1F4dWfy_BCA5koUZwrCGC-xxOnxmymcwR?usp=sharing) to show the changes working with a custom backend, for demo purposes I used XLA with a TPU machine as this is publicly available. The notebook runtime has our fork pointed to this branch, with an XLA-specific device specification file. I run `pytest` for just `test_generate_fp16` in `test_modeling_opt.py`. 

XLA does not support FP16, but we just want confirmation that the test is running on the correct device, you can take a look at the pytest output in the notebook for evidence, which shows the backend-specific error:
```
RuntimeError: torch_xla/csrc/convert_ops.cpp:86 : Unsupported XLA type 10
```

**Comment by arsalanu** [comment]
Hi @ydshieh, I have a [Colab notebook](https://colab.research.google.com/drive/1F4dWfy_BCA5koUZwrCGC-xxOnxmymcwR?usp=sharing) to show the changes working with a custom backend, for demo purposes I used XLA with a TPU machine as this is publicly available. The notebook runtime has our fork pointed to this branch, with an XLA-specific device specification file. I run `pytest` for just `test_generate_fp16` in `test_modeling_opt.py`. 

The XLA backend fails with an op support issue, but we just want confirmation that the test is running on the specified device, you can take a look at the pytest output in the notebook for evidence, which shows the backend-specific error:
```
RuntimeError: torch_xla/csrc/convert_ops.cpp:86 : Unsupported XLA type 10
```

**Comment by ydshieh** [comment]
Could you share the content of `spec.py` you used in colab? Thank you

**Comment by arsalanu** [comment]
Sure! This is what `spec.py` looks like for the example:

```
import torch
import torch_xla
import torch_xla.core.xla_model as xm

DEVICE_NAME=str(xm.xla_device())

MANUAL_SEED_FN = xm.set_rng_state
DEVICE_COUNT_FN = None
EMPTY_CACHE_FN = None
```

I believe the contents should be visible in the notebook outputs as well as I `cat` the `spec.py` file in the cell just before I run `pytest` (second-last cell)

**Comment by ydshieh** [comment]
Hi @arsalanu . I haven't been able to try with TPU, as it always complains about no backend available. It doesn't matter much.
I see this PR is on top of a very old commit on the main branch. Could you pull the latest main and rebase this PR on top of the latest main?

I will request a review from a core maintainer tomorrow. Thank you in advance!

File: docs/source/en/testing.md
Comment (statelesshz) [reply docs/source/en/testing.md:539]: The equivalent API in NPU is `torch.npu.manual_seed` or `torch.npu.manual_seed`. I prefer to use `torch.npu.manual_seed` as it's more similar to the API we used on GPU.

**Review by ji-huazhong — CHANGES_REQUESTED** [review]
Just some small suggestion about the doc. Otherwise all looks good!

**Comment by statelesshz** [comment]
Hi @ydshieh I've tested this amazing work on Ascend NPU. Just refer to the Colab notebook https://github.com/huggingface/transformers/pull/25870#issuecomment-1747121445 shared by @arsalanu for testing.  Here is the `spec.py`  I used:
```
import torch
import torch_npu
# !! Further additional imports can be added here !!
# Specify the device name (eg. 'cuda', 'cpu', 'npu')
DEVICE_NAME = 'npu'
# Specify device-specific backends to dispatch to.
# If not specified, will fallback to 'default' in 'testing_utils.py`
MANUAL_SEED_FN = torch.npu.manual_seed_all
EMPTY_CACHE_FN = torch.npu.empty_cache
DEVICE_COUNT_FN = torch.npu.device_count
```

And when I do the following test instruction:
```
TRANSFORMERS_TEST_BACKEND="torch_npu" TRANSFORMERS_TEST_DEVICE="npu"  TRANSFORMERS_TEST_DEVICE_SPEC="spec.py" python3 -m pytest tests/models/opt/test_modeling_opt.py -k "test_generate_fp16"
```

It complains a `Unknown testing device` error
```
ImportError while loading conftest '/data/hf_test/transformers/conftest.py'.
conftest.py:25: in <module>
    from transformers.testing_utils import HfDoctestModule, HfDocTestParser
src/transformers/testing_utils.py:724: in <module>
    raise RuntimeError(
E   RuntimeError: Unknown testing device specified by environment variable `TRANSFORMERS_TEST_DEVICE`: npu
```
This is because we need to register test backend first before using test device of third-party accelerators, so I make a small change, see https://github.com/huggingface/transformers/pull/26708
and everything is fine now :-)

The output log is as following：
```
(hf_test) [root@localhost transformers]# TRANSFORMERS_TEST_BACKEND="torch_npu"  TRANSFORMERS_TEST_DEVICE="npu" TRANSFORMERS_TEST_DEVICE_SPEC="spec.py" python3 -m pytest tests/models/opt/test_modeling_opt.py -k "test_generate_fp16"
=================================================== test session starts ====================================================
platform linux -- Python 3.8.18, pytest-7.4.2, pluggy-1.3.0
rootdir: /data/hf_test/transformers
configfile: setup.cfg
plugins: xdist-3.3.1, timeout-2.2.0
collected 124 items / 123 deselected / 1 selected                                                                          

tests/models/opt/test_modeling_opt.py .                                                                              [100%]

===================================================== warnings summary =====================================================
../../anaconda/envs/hf_test/lib/python3.8/site-packages/_pytest/config/__init__.py:1373
  /data/anaconda/envs/hf_test/lib/python3.8/site-packages/_pytest/config/__init__.py:1373: PytestConfigWarning: Unknown config option: doctest_glob
  
    self._warn_or_fail_if_strict(f"Unknown config option: {key}\n")

tests/test_modeling_common.py:2746
  /data/hf_test/transformers/tests/test_modeling_common.py:2746: PytestUnknownMarkWarning: Unknown pytest.mark.flash_attn_test - is this a typo?  You can register custom marks to avoid this warning - for details, see https://docs.pytest.org/en/stable/how-to/mark.html
    @mark.flash_attn_test

tests/test_modeling_common.py:2773
  /data/hf_test/transformers/tests/test_modeling_common.py:2773: PytestUnknownMarkWarning: Unknown pytest.mark.flash_attn_test - is this a typo?  You can register custom marks to avoid this warning - for details, see https://docs.pytest.org/en/stable/how-to/mark.html
    @mark.flash_attn_test

tests/test_modeling_common.py:2815
  /data/hf_test/transformers/tests/test_modeling_common.py:2815: PytestUnknownMarkWarning: Unknown pytest.mark.flash_attn_test - is this a typo?  You can register custom marks to avoid this warning - for details, see https://docs.pytest.org/en/stable/how-to/mark.html
    @mark.flash_attn_test

tests/test_modeling_common.py:2857
  /data/hf_test/transformers/tests/test_modeling_common.py:2857: PytestUnknownMarkWarning: Unknown pytest.mark.flash_attn_test - is this a typo?  You can register custom marks to avoid this warning - for details, see https://docs.pytest.org/en/stable/how-to/mark.html
    @mark.flash_attn_test

tests/test_modeling_common.py:2894
  /data/hf_test/transformers/tests/test_modeling_common.py:2894: PytestUnknownMarkWarning: Unknown pytest.mark.flash_attn_test - is this a typo?  You can register custom marks to avoid this warning - for details, see https://docs.pytest.org/en/stable/how-to/mark.html
    @mark.flash_attn_test

tests/test_modeling_common.py:2931
  /data/hf_test/transformers/tests/test_modeling_common.py:2931: PytestUnknownMarkWarning: Unknown pytest.mark.flash_attn_test - is this a typo?  You can register custom marks to avoid this warning - for details, see https://docs.pytest.org/en/stable/how-to/mark.html
    @mark.flash_attn_test

../../anaconda/envs/hf_test/lib/python3.8/site-packages/torch/utils/cpp_extension.py:28
  /data/anaconda/envs/hf_test/lib/python3.8/site-packages/torch/utils/cpp_extension.py:28: DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
    from pkg_resources import packaging  # type: ignore[attr-defined]

../../anaconda/envs/hf_test/lib/python3.8/site-packages/pkg_resources/__init__.py:2871
  /data/anaconda/envs/hf_test/lib/python3.8/site-packages/pkg_resources/__init__.py:2871: DeprecationWarning: Deprecated call to `pkg_resources.declare_namespace('google')`.
  Implementing implicit namespace packages (as specified in PEP 420) is preferred to `pkg_resources.declare_namespace`. See https://setuptools.pypa.io/en/latest/references/keywords.html#keyword-namespace-packages
    declare_namespace(pkg)

tests/models/opt/test_modeling_opt.py::OPTModelTest::test_generate_fp16
  /data/hf_test/transformers/tests/models/opt/test_modeling_opt.py:116: UserWarning: AutoNonVariableTypeMode is deprecated and will be removed in 1.10 release. For kernel implementations please use AutoDispatchBelowADInplaceOrView instead, If you are looking for a user facing API to enable running your inference-only workload, please use c10::InferenceMode. Using AutoDispatchBelowADInplaceOrView in user code is under risk of producing silent wrong result in some edge cases. See Note [AutoDispatchBelowAutograd] for more details. (Triggered internally at /opt/_internal/cpython-3.8.17/lib/python3.8/site-packages/torch/include/ATen/core/LegacyTypeDispatch.h:74.)
    input_ids[:, -1] = self.eos_token_id  # Eos Token

tests/models/opt/test_modeling_opt.py::OPTModelTest::test_generate_fp16
  /data/hf_test/transformers/src/transformers/generation/utils.py:1260: UserWarning: Using the model-agnostic default `max_length` (=20) to control the generation length. We recommend setting `max_new_tokens` to control the maximum length of the generation.
    warnings.warn(

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
===================================== 1 passed, 123 deselected, 11 warnings in 15.02s =====================================
```

**Comment by statelesshz** [comment]
Hi @arsalanu, please take a look at my result https://github.com/huggingface/transformers/pull/25870#issuecomment-1754523487 thx -:)

File: docs/source/en/testing.md
Comment (ydshieh) [reply docs/source/en/testing.md:527]: BTW, what's the difference between `torch_npu` and `torch.npu`. I had difficult time to find `torch.npu`, but even `torch_npu` I can't find its doc easily.

**Comment by arsalanu** [comment]
Hi, thanks for your comments @statelesshz, @ydshieh  and apologies for the delay, I was on vacation last week.

I have made the requested changes to the docs to reflect the usage of the `npu` backend and rebased on main, and thank you for testing it. It makes sense to register and import the backend before using the device. #26708 LGTM 👍 

I have not yet copied the changes from that PR into this one, if you confirm its OK with you I can make your changes here.

File: docs/source/en/testing.md
Comment (statelesshz) [reply docs/source/en/testing.md]: 1. `torch_npu` is an officially recognized pytorch integration plugin to support ascend npu using the pytorch framework. Ref: [Improved third-party device support](https://pytorch.org/blog/pytorch-2-1/).
2. `torch_npu` is the python package name, and relevant documentation can be obtained [here](https://github.com/ascend/pytorch).
3. The API currently supported by `torch_npu` can be obtained from [here](https://github.com/Ascend/pytorch/blob/master/docs/api/torch_npu_apis.md).
The cuda-related interfaces prefixed with `torch.cuda` correspond to the npu-related interfaces prefixed with `torch.npu`. When using them, you only need to replace `torch.c
uda.xxx` with `torch.npu.xxx`.
Take `torch.cuda.set_device` as an example:
```python
...

# torch.cuda.set_device("cuda:0")
torch.npu.set_device("npu:0")

...
```

@ydshieh Hope the above explanation can solve your confusion. Apologies for the delay :(

**Comment by ydshieh** [comment]
I mean for @arsalanu to include your work - typo 😅

**Comment by arsalanu** [comment]
I will manually include the change, I believe when this is merged we can add you @statelesshz as a contributer in this PR? @ydshieh I am not sure but are you able to do this?

**Comment by ydshieh** [comment]
Yes, I only need to add an entry in the merge message when I click the merge button.

**Comment by arsalanu** [comment]
Thanks, @statelesshz - I've made your changes to this PR. 

@ydshieh when you can, could you please change this PR from a 'draft' to ready-for-review if you feel it is ready to be approved? (I don't have the access to do this) Thank you!

**Comment by ydshieh** [comment]
Sure. BTW, could you run `make style` and/or `make quality` to fix the code quality issue.
Or more simply,
```
black --check examples tests src utils
```

**Comment by arsalanu** [comment]
Looks like they're all passing now 👍

**Comment by ydshieh** [comment]
Hi @LysandreJik 

This PR is ready for a review from core maintainers 🙏 . It enables the testing on different type of accelerators. 

The only thing I see that might be a bit inconvenience is the file `testing_utils.py` might get (even) larger if we ever need to add more device specific methods: see below

If this is the case, we can move the actual definitions to a new file and import them from there to `testing_utils`.

https://github.com/huggingface/transformers/blob/a1230d45c2a5e9ef89c832121273b780b1be7c3a/src/transformers/testing_utils.py#L2201-L2281

**Comment by statelesshz** [comment]
looking forward to this  PR being accepted🤗

**Review by ydshieh — APPROVED** [review]

**Review by LysandreJik — COMMENTED** [review]
@fxmarty, given your work with other accelerators, do you have any feedback on this PR?

File: tests/models/opt/test_modeling_opt.py
Comment (fxmarty) [inline tests/models/opt/test_modeling_opt.py:295]: This test should probably just be skipped if the device does not support fp16 (alternatively adding `@require_torch_fp16`).

**Review by fxmarty — APPROVED** [review]
LGTM

**Review by LysandreJik — APPROVED** [review]
After fixing @fxmarty's feedback, LGTM & merge ahead

File: tests/models/opt/test_modeling_opt.py
Comment (statelesshz) [reply tests/models/opt/test_modeling_opt.py:295]: cc @arsalanu

**Comment by arsalanu** [comment]
Thank you @LysandreJik, @fxmarty for the look-over. I'm just waiting for the tests to pass and then will merge.

**Comment by arsalanu** [comment]
@ydshieh (or @vvvm23!) I don't have the access to merge this. It looks ready to go, could one of you click the button? 

Reminder to include @statelesshz as a contributer :)

**Comment by vvvm23** [comment]
I don't have the ability sadly 😅

**Comment by ydshieh** [comment]
Thank you again @vvvm23 @arsalanu and @statelesshz 

(and especially reminding me to add @statelesshz as coauthor!)

**Comment by vvvm23** [comment]
image
Email for @statelesshz seems right but name is wrong in the commit 🤔

**Comment by ydshieh** [comment]
My bad 😭 . But I think GitHub counts with the email, so @statelesshz still appears

Screenshot 2023-10-24 202831

**Pull request merged**
